test(pack): add `vim.pack` tests

echasnovski opened the pull request:
Cover `vim.pack` with tests.

Resolve #34767

---

It is currently only in the state to show and discuss the overall approach to testing. TL;DR: deliberately create Git repos with specific tested properties during test execution and use `file://...` as `src`. Ideally those repos should be created as infrequently as possible (i.e. not in `before_each()` and cleaned in `after_each()`), but I'd need some help with that.

cc: @justinmk, @lewis6991

justinmk commented on test/functional/pack/testutil.lua:
this can all just live in `test/functional/plugin/pack_spec.lua` , we don't need an entire `pack/` namespace for 1 plugin.

echasnovski commented:
> Are you aware of `setup()` and `teardown()` (they are the "once per file" variant of before_each/after_each).

Yes, I tried them first, but `make functionaltest` errored because it could not find callable `vim.system` (it was executed in test runner, not child tested process). There is s a TODO in the `get()` test with more details.

> LGTM but I hope all the tests can live in pack_spec.lua, ...

The reason was that it requires non-trivial amount of helpers. As far as I can see, the pattern is to extract that into a separate 'testutil.lua' file.

> ... and (hope) we don't need 10k lines of tests.

10k is too much, but proper coverage of intended use cases might get verbose indeed. We'll see.

echasnovski commented:
> > but `make functionaltest` errored because it could not find callable `vim.system` (it was executed in test runner, not child tested process).
> 
> Call `n.clear()` in setup() ?

Yes, indeed, that does work. My bad. Thanks! I guess the name `clear()` was what confused me (i.e. I thought child process should be already set up and `n.clear()` only makes it ... clear).

echasnovski commented on test/functional/plugin/pack_spec.lua:
```diff
@@ -1,40 +1,369 @@
-      -- TODO
-      -- Should handle (not let it terminate the function) and report errors from pack_add()
+    it('can install from the Internet', function()
+      t.skip(fn.getenv('TEST_INTEG') == vim.NIL, 'Integration test: requires Internet connection')
+      exec_lua(function()
+        vim.pack.add({ 'https://github.com/neovim/nvim-lspconfig' })
+      end)
+      eq(true, exec_lua('return pcall(require, "lspconfig")'))
```
Added a separate test for `vim.pack.add()` to be able to install from the Internet. As suggested [here](https://github.com/neovim/neovim/issues/34767#issuecomment-3036644145), this is done behind a special `TEST_INTEG` environment variable, which is enabled on CI. Seems to be working.

justinmk commented:
> I thought child process should be already set up and `n.clear()` only makes it ... clear).

I'd be in favor of renaming it, I didn't have the courage because of the churn.

justinmk commented on .github/workflows/test.yml:
```diff
@@ -25,6 +25,7 @@ env:
   VALGRIND_LOG: ${{ github.workspace }}/build/log/valgrind-%p.log
   # TEST_FILE: test/functional/core/startup_spec.lua
   # TEST_FILTER: foo
+  TEST_INTEG: true
```
I think only 1 CI job should enable this. Also, `NVIM_TEST_INTEG` was introduced [recently](https://github.com/neovim/neovim/blob/8a7bfdc0792229e87fcaebdcf2e877fb91e2455c/test/README.md?plain=1#L524) (but not enabled in CI), so let's use that (I don't have a preference on the name though).

echasnovski replied:
I'd think that at least testing on sufficiently different environments is worth it. For example: basic (Linux+LuaJIT), MacOS, Windows, PUC Lua. This would require some magic in yaml files, though. Does that sound good?

echasnovski replied:
[Enabled on release build](https://github.com/neovim/neovim/pull/35052/commits/9b486f6f55538a20021c5e8b81d54c576210365c).

justinmk commented on runtime/lua/vim/pack.lua:
```diff
@@ -209,7 +209,11 @@ end
+  return type(x) == 'string' or (type(x) == 'table' and pcall(x.has, x, '1'))
+end
+
+local function is_nonempty_string(x)
```
usually call this `is_blank` (or `not is_blank`). we probably should add `vim.text.isblank`

echasnovski replied:
Blank test/line/string usually means that it consists from zero or more whitespace characters. I.e. empty line is blank line, but not every blank line is empty. It is like this both in Vim (like `:h [:blank:]`, `:h prevnonblank`, or in text like "Put cursor at first non-blank in the line") and in other languages.

Here I think it only makes sense to not allow empty string (because it conflicts with file system actions), while only whitespace characters *technically* should work.

justinmk reviewed (approved):
LGTM after comments

It looks like the plugin is writing to stdout and this is showing up in the test report? That's unusual, and definitely something we want to avoid, it makes the test logs hard to read.

```
2025-07-27T17:21:40.3160527Z [32m[2mRUN     [0m[0m T4867 vim.pack add() recognizes several `version` types: (vim.pack)   0%: Installing plugins (0/3)
2025-07-27T17:21:40.3161096Z (vim.pack)  33%: Installing plugins (1/3) - defbranch
2025-07-27T17:21:40.3161408Z (vim.pack)  66%: Installing plugins (2/3) - basic
2025-07-27T17:21:40.3161683Z (vim.pack) 100%: Installing plugins (3/3) - semver
2025-07-27T17:21:40.3161987Z [2m62.00 ms[0m [1m[32mOK[0m[0m
```

> I thought child process should be already set up and `n.clear()` only makes it ... clear).

I'd be in favor of renaming it, I didn't have the courage because of the churn.

echasnovski committed:
fix(pack): use 'coxpcall.lua' on non-LuaJIT

Problem: `attempt to yield across metamethod/C-call boundary` error when
  trying to use `vim.pack.add()`.

Solution: use `pcall()` variant from 'coxpcall' on non-LuaJIT version of
  Lua.

echasnovski committed:
fix(pack): improve `vim.pack.add()` input validation

echasnovski committed:
fix(pack): ignore 'loadplugins' when sourcing 'after/plugin' scripts

Problem: Sourcing of 'after/plugin' scripts depends on the value of
  'loadplugins' option. It is redundant, as it has effect only during
  startup, while it is combined with `vim.v.vim_did_enter == 1` (i.e.
  "it is after starting up") condition.

Solution: Ignore it.

echasnovski committed:
feat(pack): use colored `nvim_echo` chunks to show progress report

Problem: using `print()` to show progress report writes to `stdout` when
  in `--headless` mode (interferes with the testing output) and doesn't
  allow coloring.

Solution: use `nvim_echo` with colored chunks.

echasnovski committed:
fix(pack): validate installed plugins in `update()` and `del()`

Problem: Currently `update()` and `del()` silently ignore input plugin
  names that are not for already installed plugin. This might lead to
  confusion because they are not explicitly reported.

Solution: Check that all input names are for installed plugins and error
  otherwise.

justinmk commented on runtime/lua/vim/pack.lua:
```diff
@@ -367,13 +382,16 @@ local function new_progress_report(title)
     local progress = kind == 'end' and 'done' or ('%3d%%'):format(percent)
-    print(('(vim.pack) %s: %s %s'):format(progress, title, fmt:format(...)))
+    local details = (': %s %s'):format(title, fmt:format(...))
+    local chunks = { { '(vim.pack)', 'ModeMsg' }, { ' ' }, { progress, 'WarningMsg' }, { details } }
```
Would prefer to use `vim.pack:` (or just `pack:`) as the message prefix. That is the conventional message format. Don't want new features to start copying a different `(foo)` convention.

echasnovski commented:
Updated formatting of progress messages to use `vim.pack:` as prefix. Also dropped `:` after percent of progress since it feels awkward with two `:`. It now looks like this:

<img width="2880" height="1734" alt="screenshot_2025-07-30_21:11:41" src="https://github.com/user-attachments/assets/83836ad2-be2d-4b67-af11-31f793500864" />

Also:
- Changed how newer versions are suggested (based purely on semver comparison instead of from Git history).
- I think I found happy medium for testing overall look of confirmation buffer through screenshot. Although highlighting testing is more explicit, it required some manipulation to hide dynamically constructed paths (both for source repo and installed plugins) and commit hashes (as they are cleanly created for each file testing). I though about not showing full paths in confirmation buffer, but I think it is important in order to be able to `gf` and `gx` into file path or link.

    Let me know if I did something wrong and there is a more preferred way to do screenshot testing. I do not plan to add more here and make the rest of the tests through text comparison.

justinmk reviewed (approved):
Looks great!

echasnovski committed:
fix(pack): use `vim.pack:` as message/progress prefix

echasnovski committed:
fix(pack): ensure explicit default `version` in events (where possible)

Problem: Both `PackChangedPre` and `PackChanged` contain |event-data|
  with plugin's `spec`. It looks like a good idea to have all its
  triggers contain the same format across all kinds ("install",
  "update", "delete"). There are several choices:
    - Have it be as verbatim as supplied to `vim.pack.add()`, i.e. can
      be either string or table. A bit too ambiguous.
    - Have it be table with `src` and `name` inferred. This requires
      less work for "install", but more work for "update" and "delete"
      (since they use `vim.pack.get()` which already infers default
      `version`).
    - Have it be table with *all* defaults made explicit. This looks
      like the best approach, but requires extra care to only infer
      default `version` when needed (i.e. avoid inferring during regular
      load) because it is costly in terms of startup time.
      This might also introduce inconsistency when dealing with
      lockfile(s) as information there should be as close to what user
      supplied as possible. Address that when dealing with lockfile.

Solution: Ensure explicit `version` in all events where possible.

echasnovski committed:
fix(pack): make newer version suggestions more robust

Problem: New version suggestions in update confirmation buffer might
  include semver tags that were committed later but for versions that
  are not greater than current. Like if versions committed in order
  `v0.2.0` - `v0.3.0` - `v0.2.1` - `v0.3.1`, then when on `v0.3.0` both
  `v0.2.1` and `v0.3.1` are suggested, but only the latter is newer as
  a version. This is because those tags are computed with post-processed
  `git tag --list --contains HEAD`.

Solution: Compute all semver tags and filter only those greater than the
  latest version available at HEAD.

echasnovski committed:
fix(pack): ensure consistent order of changelog during update

Problem: Left (`<` prefix) and right (`>` prefix) parts of update
  changelog come in not fixed order: it can be left-right or right-left.

Solution: Ensure left-right order.

echasnovski committed:
fix(pack): open confirmation buffer in tabpage after the current one

echasnovski committed:
ci: enable integration tests on Release builds

echasnovski committed:
test(pack): add `vim.pack` tests

echasnovski committed:
refactor(pack): use `vim.cmd` methods with structured input

justinmk commented:
@echasnovski any idea why this test always fails on windows ci:

```
ERROR    test/functional\plugin\pack_spec.lua @ 388: vim.pack add() startup works in init.lua
test\functional\testnvim.lua:139: Lua: module 'plugindirs' not found:
	no field package.preload['plugindirs']
	no file '.\plugindirs.lua'
	no file 'D:\a\neovim\neovim\build\bin\lua\plugindirs.lua'
	no file 'D:\a\neovim\neovim\build\bin\lua\plugindirs\init.lua'
	no file '.\plugindirs.dll'
	no file 'D:\a\neovim\neovim\build\bin\plugindirs.dll'
	no file 'D:\a\neovim\neovim\build\bin\loadall.dll'
```
